drivers/usb/udc/udc_stm32.c

```diff
@@ -844,6 +844,15 @@ static int udc_stm32_ep_set_halt(const struct device *dev,
 		return -EIO;
 	}
 
+	/*
+	 * Mark EP as halted only if it isn't a control endpoint.
+	 * Hardware will clear halt bit of EP0 automatically such
+	 * that software never sees control endpoints as halted.
```

Comment:
I think that is not quite correct, and I suggest removing it. For example, if the controller is DWC2, then the controller writes received setup packet to the FIFO regardless of NAK, STALL, or available FIFO size. It is up to the driver, then, to handle it properly, see `dwc2_handle_evt_setup()` in udc_dwc2.c.

Reply:
@jfischer-no Don't hesitate to put NACK if you really mean to. Otherwise I take it as a non blocking comment.